### PR TITLE
Add TFSpeech2Text

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR adds a TF port of Speech2Text. A summary of the changes:
- This model borrows a lot of code from TFBart, just like Speech2Text borrowed from Bart;
- Tried to follow the changes in other PRs to enable smooth interoperation with other parts of transformers (e.g. auto classes), might be missing a few things 👼 ;
- This seems to be the first TF model with speech as input, so I had to touch common TF code to enable correct data piping and misc operations (e.g. enable loading Conv1D PT weights into TF);
- Likewise, there were a few tests in `test_modelling_tf_common.py` that didn't quite fit this new kind of model.

TODO: 
- [x] create TF version of the weights, so we can load a model without `from_pt=True`

### CODE REVIEW

**Comment by gante** [comment]
Tagging @sgugger as a core dev, @patil-suraj as a core dev + original creator of our `Speech2Text`, and @Rocketknight1 as the TensorFlow boy. Feel free to redirect the reviews if you know of better people to review.

Some pipeline tests are failing almost surely due to the changes in `generation_tf_utils.py`, as some models expect `(encoder_outputs, past)` in the `past` variable and others don't -- having a look, but open to suggestions.

File: src/transformers/generation_tf_utils.py
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:880]: Let's double check with @patrickvonplaten for this change.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (sgugger) [inline src/transformers/modeling_tf_pytorch_utils.py:213]: Very clean!

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:234]: We'll need to fix this assert in the original model (no new assert in the modeling files normally, but this is a copied from).

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:475]: All the shapes should be in code blocks, not italics.

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:531]: Night, this may fit in one line.

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:1091]: This is hard to read with the formatting, so put this in a separate line.

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:1548]: Please use a test and raise the appropriate Error :-)

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (sgugger) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:1572]: Same here.

File: tests/test_modeling_tf_common.py
Comment (sgugger) [inline tests/test_modeling_tf_common.py:463]: Is there another (more general) check we could use here?

File: tests/test_modeling_tf_common.py
Comment (sgugger) [inline tests/test_modeling_tf_common.py:1017]: For those two tests, maybe we should overwrite the corresponding tests in the speech2text test file.

**Review by sgugger — APPROVED** [review]
Very nice addition, great work!

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py]: Excellent point, for some reason I didn't notice that it was in the keyword arguments (and thus the current version would fail) 👍  Checking against `None` now

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py]: I've reworked this part to be less intrusive: the `past` variable is not touched, `encoder_outputs` is passed to `prepare_inputs_for_generation` a few lines below. This means that it will be ignored as part of `kwargs` in most models, and encoder-decoder models can now explicitly access it if they wish to do so (like `TFSpeech2Text`).

This change also solves the `assert` comments below, in `prepare_inputs_for_generation`.

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (gante) [reply src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py]: Copied the PT check (that raises a `ValueError`) to both models

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (gante) [reply src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py]: Copy-pasta from TFBart, changed in both 👍

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py
Comment (gante) [reply src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py]: Well, it turns out that many models used a copy of this 😅  Changed them all 👍

File: tests/test_modeling_tf_common.py
Comment (gante) [reply tests/test_modeling_tf_common.py:463]: By more general do you mean instead of hardcoding the possible names in `model_class.__name__`? We have `TF_MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING_NAMES`, but it doesn't cover `TFSpeech2TextModel` :(

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:763]: Is this only needed for the new model or was it a bug to not have done this previously?

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:884]: If past and encoder_outputs is the same, why do we pass both here? Because the name `enocder_outputs` is cleaner? Would it maybe be better to keep `past` for now for consistency?

File: tests/test_modeling_tf_speech_to_text.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_speech_to_text.py:577]: Let's upload the models with TF weights no?

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:763]: I'm a bit worried about this as it could lead to some unexpected errors in the more complicated models like TFRag. Could we maybe run the following tests here to double check that this change is ok:

`RUN_SLOW=1 pytest tests/test_modeling_tf_bart.py`
`RUN_SLOW=1 pytest tests/test_modeling_tf_t5.py`
`RUN_SLOW=1 pytest tests/test_modeling_tf_rag.py` (not that you need some more dependencies to be installed here)

**Review by patrickvonplaten — COMMENTED** [review]
Looks good to me in general! Maybe just two final things before merging:

- Quickly run some slow tests to make sure we haven't broken `generate` accidently for some of the more complicated models
- Should we maybe leave `past` for now until we improve the naming of generate more globally?

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py]: Spot on, it failed for those models. 

I've reworked these lines to behave like pytorch's `_expand_inputs_for_generation()`, which is used in all modalities. I've tested the changes against bart, bert, t5, vit, and s2t -- let me know what you think, @patrickvonplaten

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py:866]: Once again, your comment led to insightful changes -- digging deeper, I found that the `past_key_values` output of the decoder is different on PT and TF (TF's is a tuple with the encoder's hidden states and past key values, and PT's is only the later). I had adapted the code from PT, so my `past_key_values` was missing the encoder outputs, and I had to pass them explicitly in order to work.

I've reverted this change and made TFSpeech2Text's `past_key_values` match the expected TF shape. An updated TF generate would probably have prevented this error 🤔

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py]: Cool! Could you please also run the slow `TFRag` tests - it's a bit annoying as it's a huge and slow test, but it's the model that is most prone to break if there are changes in tf generate

Comment (gante) [reply src/transformers/generation_tf_utils.py]: Can confirm it passes TFRag tests as well (had to spin up a larger machine 😅 )

**Comment by Rocketknight1** [comment]
Cloned your branch and did some experimentation and LGTM! I noticed one issue - `model.save()` seems to encounter problems, but I'm not totally sure of why, or whether it's limited to this model or not. `save_pretrained` and `save_weights` worked correctly, and saving in `SavedModel` format has always been a bit shaky for us, so this isn't critical.

**Review by Rocketknight1 — APPROVED** [review]

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py]: Cool! Good for me then

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1122]: can we delete this?

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:723]: Slightly worried here about the TF vision tests. Can you check this one as well:

```
RUN_SLOW=1 pytest tests/test_modeling_tf_vision_encoder_decoder.py
```

**Review by patrickvonplaten — COMMENTED** [review]
One final vision tests as mentioned before and then it's good for me as well

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py:723]: Can confirm that they pass 👍

File: src/transformers/generation_tf_utils.py
Comment (gante) [reply src/transformers/generation_tf_utils.py:1122]: 🤦

File: src/transformers/models/speech_to_text/modeling_speech_to_text.py
Comment (patrickvonplaten) [inline src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: @patil-suraj - is this change ok? Why exactly do we have to do this here actually?

Comment (patrickvonplaten) [reply src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: I'm not sure we can just save a tensor like this in PyTorch. When doing `model.to(device)` does this tensor correctly move to the device as well?

Comment (patrickvonplaten) [reply src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: I don't really understand why this would cause an issue in TF<=>PT exactly - what was the issue? BTW, we always wrap the static sinusoidal encodings in a `nn.Parameter()` or `nn.Embedding(...)`, E.g. Vit's positional embeddings, DistilBERT's position embeddings. Why should we do it differently here?

Comment (gante) [reply src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: > what was the issue?

This is the only model that I could find that may change the size of the embedding layer in the forward call ([here](https://github.com/huggingface/transformers/blob/master/src/transformers/models/speech_to_text/modeling_speech_to_text.py#L188)). The TF version for other embeddings allocates these named model parameters in [`build()`](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#build), which is expected to happen before the first `call()`. Since `nn.Parameter()` is mostly a wrapper to add the variable to the model's list of parameters (AFAIK), then dropping it means we don't need a named parameter in TF, which means we can simplify the code and avoid solving the resizing issue.

Comment (sgugger) [reply src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: Technically, someone who fine-tuned (or just saved and reshared) a Speecj2Text model will have this weight in the checkpoint. I don't know how it's used exactly, so I don't know if it's breaking to ignore it (we would have to put it in the `_keys_to_ignore_on_load_unexpected`).

Comment (gante) [reply src/transformers/models/speech_to_text/modeling_speech_to_text.py:156]: (sorted on Slack -- conclusion: reverting this change since it is needed for the `to` method)

**Comment by gante** [comment]
@patrickvonplaten reverted the previous change and added the embedding weights as named variables in TF

**Review by patrickvonplaten — APPROVED** [review]
Thanks for working on it!

**Pull request merged**
